docs/advanced/UsageWithTypescript.md

```diff
@@ -0,0 +1,219 @@
+
+```ts
+// src/store/chat/types.ts
+
+export enum ChatActions {
+  SendMessage = '@CHAT:SEND/MESSAGE'
+}
+
+export interface ISendMessageAction {
+  type: ChatActions.SendMessage
+  payload: IMessage
+}
+```
+
+With these types we can now also type check chat's action creators:
+
+```ts
+// src/store/chat/actions.ts
+
+import { IMessage, ISendMessageAction, ChatActions } from './types'
+
+export type sendMessageType = (newMessage: IMessage) => ISendMessageAction
+
+export const sendMessage: sendMessageType = newMessage => ({
+  type: ChatActions.SendMessage,
+  payload: newMessage
+})
+```
+
+System Action Constants and Actions:
+
+```ts
+// src/store/system/types.ts
+
+export enum SystemActions {
+  UpdateSession = '@SYSTEM:UPDATE/SESSION'
+}
+
+export interface IUpdateSessionAction {
+  type: SystemActions.UpdateSession
+  payload: ISystemState
+}
+```
+
+With these types we can now also type check system's action creators:
+
+```ts
+// src/store/system/actions.ts
+
+import { SystemActions, IUpdateSessionAction, ISystemState } from './types'
+
+export type updateSessionType = (
+  newSession: ISystemState
+) => IUpdateSessionAction
+
+export const updateSession: updateSessionType = newSession => ({
+  type: SystemActions.UpdateSession,
+  payload: newSession
+})
+```
+
+## Type Checking Reducers
+
+Reducers are just pure functions that take the previous state, an action and then return the next state. Redux provides a `Reducer` type which can be imported to assist with type checking.
+
+Type checked chat reducer:
+
+```ts
+// src/store/chat/reducers.ts
+
+import { Reducer } from 'redux'
+import { IChatState, ChatActions } from './types'
+
+const initialState: IChatState = {
+  messages: []
+}
+
+export const chatReducer: Reducer<IChatState> = (
+  state = initialState,
+  action
+) => {
+  switch (action.type) {
+    case ChatActions.SendMessage: {
+      return {
+        messages: [...state.messages, action.payload]
+      }
+    }
+    default:
+      return state
+  }
+}
+```
+
+Type checked system reducer:
+
+```ts
+// src/store/system/reducers.ts
+
+import { Reducer } from 'redux'
+import { SystemActions, ISystemState } from './types'
+
+const initialState: ISystemState = {
+  loggedIn: false,
+  session: '',
+  userName: ''
+}
+
+export const systemReducer: Reducer<ISystemState> = (
+  state = initialState,
+  action
+) => {
+  switch (action.type) {
+    case SystemActions.UpdateSession: {
+      return {
+        ...state,
+        ...action.payload
+      }
+    }
+    default: {
+      return state
+    }
+  }
+}
+```
+
+Putting it all together in combine reducers:
+
+```ts
+// src/store/index.ts
+
+import { systemReducer } from './system/reducers'
+import { ISystemState } from './system/types'
+
+import { chatReducer } from './chat/reducers'
+import { IChatState } from './chat/types'
+
+export interface IAppState {
+  system: ISystemState
+  chat: IChatState
+}
+
+const rootReducer: Reducer<IAppState> = combineReducers({
```

Comment:
Once again, you do not need to use `Reducer<IAppState>` here.

Also, defining `IAppState` here is not actually required! you can instead do this:

```ts
const rootReducer = combineReducers({ system, chat })
export type AppState = ReturnType<typeof rootReducer>
```

This can save some hassle in larger state trees